src/Compression/Compression.cs

```diff
@@ -169,12 +169,17 @@ private async Task FlushAsyncCore (CancellationToken cancellationToken)
 
 				// Pull out any bytes left inside deflater:
 				bool flushSuccessful;
+				EnsureCompressionMode ();
+
 				do {
 					int compressedBytes;
-					flushSuccessful = _deflater.Flush (_buffer, out compressedBytes);
-					if (flushSuccessful) {
-						await _stream.WriteAsync(_buffer, 0, compressedBytes, cancellationToken).ConfigureAwait (false);
-					}
+					// TODO Flush throws exception if _buffer is null, do we want to
+					// call EnsureBufferInitialized () to create a _buffer if it is null?
+					flushSuccessful = _deflater!.Flush (_buffer!, out compressedBytes);
+
+					EnsureNotDisposed ();
+					// TODO I think we should call EnsureBufferInitialized () for this next call as well
+					await _stream!.WriteAsync(_buffer!, 0, compressedBytes, cancellationToken).ConfigureAwait (false);
```

Comment:
@mandel-macaque Then there are some Writes and Reads that take the buffer, but once again, the buffer may be null. Should we call EnsureBufferInitialized () and give it an empty buffer?

Reply:
same as my previous comment.

Reply:
same as my previous comment.